omegaconf/built_in_resolvers.py

```diff
@@ -43,21 +129,39 @@ def env(key: str, default: Optional[str] = _DEFAULT_MARKER_) -> Optional[str]:
             raise KeyError(f"Environment variable '{key}' not found")
 
 
-def decode(expr: Optional[str], _parent_: Container) -> Any:
-    """
-    Parse and evaluate `expr` according to the `singleElement` rule of the grammar.
+# DEPRECATED: remove in 2.2
+def legacy_env(key: str, default: Optional[str] = None) -> Any:
+    warnings.warn(
+        "The `env` resolver is deprecated, see https://github.com/omry/omegaconf/issues/573",
+    )
 
-    If `expr` is `None`, then return `None`.
-    """
-    if expr is None:
-        return None
+    try:
+        return decode_primitive(os.environ[key])
+    except KeyError:
+        if default is not None:
+            return decode_primitive(default)
+        else:
+            raise ValidationError(f"Environment variable '{key}' not found")
 
-    if not isinstance(expr, str):
+
+def _get_and_validate_dict_input(
+    in_dict: Union[str, Mapping[Any, Any]],
+    root: BaseContainer,
+    resolver_name: str,
+) -> Mapping[Any, Any]:
+    if isinstance(in_dict, str):
+        # Path to an existing key in the config: use `select()`.
+        key = in_dict
+        in_dict = OmegaConf.select(
+            root, key, throw_on_missing=True, default=_DEFAULT_SELECT_MARKER_
+        )
+        if in_dict is _DEFAULT_SELECT_MARKER_:
+            raise ConfigKeyError(f"Key not found: '{key}'")
+
+    if not isinstance(in_dict, Mapping):
```

Comment:
If we limit to DictConfig nodes, don't forget to change this.

Reply:
That was done in 27e59f8